frame/bridge/ethereum/issuing/src/lib.rs

```diff
@@ -21,275 +21,347 @@
+				str::from_utf8(&register_info.symbol[..]).map_err(|_| Error::<T>::StringCF)?,
+				register_info.decimals.as_u32() as u8,
+				backing_address,
+				register_info.token_address,
+			)
+			.map_err(|_| Error::<T>::InvalidEncodeERC20)?;
 			Self::transact_mapping_factory(input)?;
-			VerifiedIssuingProof::insert(tx_index, true);
-			Self::deposit_event(RawEvent::RegisterErc20(user, backing_address, tx_index));
+			VerifiedIssuingProof::<T>::insert(tx_index, true);
+			Self::deposit_event(Event::RegisterErc20(backing_address, tx_index));
+			Ok(().into())
 		}
 
-		#[weight = <T as darwinia_evm::Config>::GasWeightMapping::gas_to_weight(0x100000)]
-		pub fn redeem_erc20(origin, proof: EthereumReceiptProofThing<T>) {
+		#[pallet::weight(<T as Config>::WeightInfo::redeem_erc20())]
+		pub fn redeem_erc20(
+			origin: OriginFor<T>,
+			proof: EthereumReceiptProofThing<T>,
+		) -> DispatchResultWithPostInfo {
 			log::debug!("start to redeem erc20 token");
-			let user = ensure_signed(origin)?;
-			let (tx_index, ethlog) = Self::verify_and_parse_proof(
-				Abi::backing_event(),
-				proof)?;
-			let backing_address = EthereumBackingAddress::get();
-			let input = Self::abi_encode_token_redeem(ethlog)?;
+			ensure_signed(origin)?;
+			let tx_index = T::EthereumRelay::gen_receipt_index(&proof);
+			ensure!(
+				!VerifiedIssuingProof::<T>::contains_key(tx_index),
+				<Error<T>>::AssetAR
+			);
+			let verified_receipt = T::EthereumRelay::verify_receipt(&proof)?;
+			let backing_address = EthereumBackingAddress::<T>::get();
+			let lock_info =
+				EthereumBacking::parse_locking_event(&verified_receipt, &backing_address)
+					.map_err(|_| Error::<T>::DecodeEventFailed)?;
+			let input = mtf::encode_cross_receive(
+				lock_info.mapped_address,
+				lock_info.recipient,
+				lock_info.amount,
+			)
+			.map_err(|_| Error::<T>::InvalidEncodeERC20)?;
 			Self::transact_mapping_factory(input)?;
-			VerifiedIssuingProof::insert(tx_index, true);
-			Self::deposit_event(RawEvent::RedeemErc20(user, backing_address, tx_index));
+			VerifiedIssuingProof::<T>::insert(tx_index, true);
+			Self::deposit_event(Event::RedeemErc20(backing_address, tx_index));
+			Ok(().into())
 		}
-	}
-}
 
-impl<T: Config> Module<T> {
-	/// The account ID of the issuing pot.
-	pub fn dvm_account_id() -> H160 {
-		let account32: AccountId32 = T::PalletId::get().into_account();
-		let account20: &[u8] = &account32.as_ref();
-		H160::from_slice(&account20[..20])
-	}
+		#[pallet::weight(<T as Config>::WeightInfo::mapping_factory_event_handle())]
+		pub fn mapping_factory_event_handle(
+			origin: OriginFor<T>,
+			input: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			let caller = ensure_signed(origin)?;
+			ensure!(input.len() >= 8, <Error<T>>::InvalidInput);
+			let factory = MappingFactoryAddress::<T>::get();
+			let factory_id = <T as darwinia_evm::Config>::AddressMapping::into_account_id(factory);
+			ensure!(factory_id == caller, <Error<T>>::AssetAR);
```

Comment:
`<Error<T>>::AssetAR)` This error is inaccurate here?  May something like `<Error<T>>::BadCaller)`